### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by elizabetereis**

### CODE REVIEW

File: include/hyro/SignalGeneratorComponent.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/** @file SignalConverterComponent.hpp
```
Comment (joaobrittoneto) [inline include/hyro/SignalGeneratorComponent.hpp]: `.h` is the Hyro convention for headers

File: include/hyro/SignalGeneratorComponent.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#include <hyro/msgs/Signal.h>
+#include <hyro/msgs/common/Basic.h>
+#include "../src/internal/SignalGenerator.hpp"
+
+namespace signal_generator
+{
+
+class SIGNAL_GENERATOR_COMPONENTS_EXPORT SignalGeneratorComponent : public hyro::Component
+{
+	public:
```
Comment (joaobrittoneto) [inline include/hyro/SignalGeneratorComponent.hpp]: Indentation used by Hyro:
zero space  for `public` and `private`
two spaces for members of class

```
class SIGNAL_GENERATOR_COMPONENTS_EXPORT SignalGeneratorComponent : public hyro::Component
{
public:

  using hyro::Component::Component;

  inline virtual
  ~SignalGeneratorComponent () override = default;

  virtual hyro::Result
  init (const hyro::ComponentConfiguration & config) override;
...
```

File: msgs/hyro/msgs/Signal.proto [hunk 3]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+
+package hyro.msgs;
+
+message Signal
+{
+	int64 timestamp = 1;
```
Comment (joaobrittoneto) [inline msgs/hyro/msgs/Signal.proto:7]: Two spaces indentation

File: src/components/DigitalConverterComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace signal_generator
+{
+
+std::shared_ptr<hyro::HyroLogger> DigitalConverterComponent::s_logger = hyro::HyroLoggerManager::CreateLogger("DigitalConverterComponent");
+
+DigitalConverterComponent::DigitalConverterComponent (const hyro::URI & uri)
+  : hyro::Component(uri)
+{
+  m_amplitude = 5.0;
```
Comment (joaobrittoneto) [inline src/components/DigitalConverterComponent.cpp]: Configure parameters in `init`, using the ComponentConfiguration argument, with a default value.

`m_amplitude = config.parameters.getParameter<double>("amplitude", 5.0);`

File: src/components/SignalGeneratorComponent.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+namespace signal_generator
+{
+
+std::shared_ptr<hyro::HyroLogger> SignalGeneratorComponent::s_logger = hyro::HyroLoggerManager::CreateLogger("SignalGeneratorComponent");
+
+SignalGeneratorComponent::SignalGeneratorComponent(const hyro::URI & uri) 
+  : hyro::Component(uri)
+{
+  m_amplitude = 5.0;
```
Comment (joaobrittoneto) [inline src/components/SignalGeneratorComponent.cpp]: Same as above. Configure values in `init` with values in `config` with default values

File: src/components/SignalGeneratorComponent.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+ *  @return Result.
+ */
+hyro::Result
+SignalGeneratorComponent::update()
+{
+  Signal signal;
+  SignalGenerator signalGen;
+
+  signal.frame_id = "Testing";
+  signal.timestamp = 1;
```
Comment (joaobrittoneto) [inline src/components/SignalGeneratorComponent.cpp]: Timestamps are really important. 
Try to use the `timestamp = hyro::TimeUtils::Now_TimeSinceEpoch();`

File: src/internal/SignalGenerator.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+/** @brief generates analog signal 
+ *  @param cosine: boolean indicating if the signal value will be sin ou cosine function 
+ *         amp: amplitude value
+ *				 step: frequency.
+ *  @return float.
+ */
+float
+SignalGenerator::getSignalValue(bool cosine, float amp, float step)
+{
+	float deg = 60.0;
```
Comment (joaobrittoneto) [inline src/internal/SignalGenerator.cpp]: It seems that you will return a constant value of `amp * cos(60 degrees)`

File: src/internal/SignalGenerator.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
+#include "SignalGenerator.hpp"
+
+namespace signal_generator
+{
+
+SignalGenerator::SignalGenerator(){
+
+}
+
+/** @brief generates analog signal
```
Comment (joaobrittoneto) [inline src/internal/SignalGenerator.cpp]: Methods descriptions just in the header file. Not in the .cpp file

File: src/internal/Thresholding.cpp [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ *  @param value: analog signal value 
+ *         amp: amplitude value
+ *				 thresh: threshold value.
+ *  @return float.
+ */
+float
+Thresholding::convertSignal(float value, float amp, float thresh){
+
+	if(value > thresh)
+		value = amp;
```
Comment (joaobrittoneto) [inline src/internal/Thresholding.cpp]: Directly return here. And two space indentation

```
if (value > thresh)
  return amp;

return -amp;

File: include/hyro/SignalGeneratorComponent.hpp (hunk 1 above)
Comment (elizabetereis) [reply include/hyro/SignalGeneratorComponent.hpp]: Done. For all.

File: include/hyro/DigitalConverterComponent.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+#include <hyro/msgs/Signal.h>
+#include "../src/internal/Thresholding.h"
+
+namespace signal_generator
+{
+
+class SIGNAL_GENERATOR_COMPONENTS_EXPORT DigitalConverterComponent : public hyro::Component{
+	
+public:
+	DigitalConverterComponent(const hyro::URI & uri);
```
Comment (joaobrittoneto) [inline include/hyro/DigitalConverterComponent.h]: two space indentation

Comment (elizabetereis) [reply include/hyro/DigitalConverterComponent.h]: done

Comment (gustavoneves12) [reply include/hyro/DigitalConverterComponent.h]: Still more than two space indentation.

We do not indent the class access modifiers (`public`, `private` and `protected`). See example below:

```c++
class DigitalConverterComponent
{
public:
  DigitalConverterComponent(const hyro::URI & uri);
}
```

File: include/hyro/SignalGeneratorComponent.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+#include <hyro/msgs/common/Basic.h>
+#include "../src/internal/SignalGenerator.h"
+
+namespace signal_generator
+{
+
+class SIGNAL_GENERATOR_COMPONENTS_EXPORT SignalGeneratorComponent : public hyro::Component
+{
+public:
+	SignalGeneratorComponent(const hyro::URI & uri);
```
Comment (gustavoneves12) [inline include/hyro/SignalGeneratorComponent.h]: Fix the indentation in your whole project.

File: src/components/DigitalConverterComponent.cpp [hunk 12]
```diff
@@ -0,0 +1,77 @@
+#include <hyro/DigitalConverterComponent.h>
+
+using namespace hyro;
+
+namespace signal_generator
+{
+
+std::shared_ptr<hyro::HyroLogger> DigitalConverterComponent::s_logger = hyro::HyroLoggerManager::CreateLogger("DigitalConverterComponent");
+
+DigitalConverterComponent::DigitalConverterComponent (const hyro::URI & uri)
```
Comment (gustavoneves12) [inline src/components/DigitalConverterComponent.cpp]: This constructor is empty. You do not need to declare the constructor here. Instead, you can replace the constructor declaration with `using hyro::Component::Component` in the headers.

File: src/components/DigitalConverterComponent.cpp [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+  auto ret = m_signals->receive(value, 1s);
+  Thresholding thresholding;
+  Signal response_signal;
+  Signal input_signal;
+  
+  if (ret == ReceiveStatus::RECEIVE_OK)
+  {
+    input_signal = *value;
+    response_signal.frame_id = "Answer";
+    response_signal.timestamp = TimeUtils::Now_TimeSinceEpoch();;
```
Comment (gustavoneves12) [inline src/components/DigitalConverterComponent.cpp:52]: Remove one **;**

File: src/components/DigitalConverterComponent.cpp [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    response_signal.value = thresholding.convertSignal(input_signal.value, m_amplitude, m_threshold);;
```
Comment (gustavoneves12) [inline src/components/DigitalConverterComponent.cpp]: Why do you use **;;**?

File: include/hyro/msgs/Signal.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+#define SIGNAL_GENERATOR_MSG_SIGNAL_H
+
+#include <signal_generator_components_export.h>
+#include <string>
+#include <hyro/common/Time.h>
+
+namespace hyro
+{
+
+struct SIGNAL_GENERATOR_COMPONENTS_EXPORT Signal
```
Comment (gustavoneves12) [inline include/hyro/msgs/Signal.h]: Two space indentation!

File: package.json [hunk 16]
```diff
@@ -0,0 +1,14 @@
+{
+	"name": "signal_generator",
```
Comment (gustavoneves12) [inline package.json:2]: Two space indentation.

File: src/components/SignalGeneratorComponent.cpp [hunk 17]
```diff
@@ -0,0 +1,127 @@
+#include <hyro/SignalGeneratorComponent.h>
+
+using namespace hyro;
+
+namespace signal_generator
+{
+
+std::shared_ptr<hyro::HyroLogger> SignalGeneratorComponent::s_logger = hyro::HyroLoggerManager::CreateLogger("SignalGeneratorComponent");
+
+SignalGeneratorComponent::SignalGeneratorComponent(const hyro::URI & uri)
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: Use `using hyro::Component::Component` in the header file.

File: src/components/SignalGeneratorComponent.cpp [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+  
+};
+
+hyro::Result
+SignalGeneratorComponent::init (const hyro::ComponentConfiguration & config)
+{
+  m_amplitude = config.parameters.getParameter<float>("amplitude", 5.0);
+  m_frequency = config.parameters.getParameter<float>("frequency", 7.0);
+  m_cosine = config.parameters.getParameter<bool>("cosine", false);
+  m_name = config.parameters.getParameter<bool>("name", "Signal");
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: The name is bool? Please fix here.

File: src/components/SignalGeneratorComponent.cpp [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+hyro::Result
+SignalGeneratorComponent::update()
+{
+  Signal signal;
+  SignalGenerator signalGen;
+
+  signal.frame_id = "Testing";
+  signal.timestamp = TimeUtils::Now_TimeSinceEpoch();;
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: Why **;;**?

File: src/internal/SignalGenerator.cpp [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+{
+
+SignalGenerator::SignalGenerator(){
+
+}
+
+float
+SignalGenerator::getSignalValue(bool cosine, float amp, float step)
+{
+	float deg = 60.0;
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp]: This variable should be a member class variable. Or if you prefer, make it static.

File: src/internal/SignalGenerator.cpp [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+	float rad = deg / 180.0 * M_PI;
+	deg += step;
+	
+	if (cosine)
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp]: Indentation always two spaces!

Comment (gustavoneves12) [reply src/internal/SignalGenerator.cpp]: Don't you like ternary operator?

File: src/internal/SignalGenerator.h [hunk 22]
```diff
@@ -0,0 +1,33 @@
+/** @file SignalGenerator.h
+ *  @brief Function prototypes for the signal generator.
+ *  @author Elizabete
+ *  @bug No known bugs.
+ */
+
+#ifndef SIGNALGENERATOR_H
+#define SIGNALGENERATOR_H
+
+#define PI 3.14159265
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.h]: You don't need to define `PI`. You can access `PI` using `M_PI`

File: tests/test_signal_generator.cpp [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+  ASSERT_EQ(Result::RESULT_OK, result);
+
+  auto inputSignal = std::make_shared<FakeInput<Signal>>("inputSignal"_uri, "api", "/generator/signals");
+  
+  ASSERT_TRUE(inputSignal->connect());
+
+  result = generator_sm.update();
+  ASSERT_EQ(Result::RESULT_OK, result);
+
+  auto valueSignal = std::shared_ptr<const Signal>();
```
Comment (gustavoneves12) [inline tests/test_signal_generator.cpp:61]: Please, check if the received values is equal the value expected.

File: tests/test_signal_generator.cpp [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+  ASSERT_TRUE(dynamic_property_access.get("amplitude", amplitude));
+  ASSERT_TRUE(dynamic_property_access.get("frequency", frequency));
+  ASSERT_TRUE(dynamic_property_access.get("cosine", cosine));
+
+  Signal signalTest;
+  outputSignal->sendAsync(signalTest);
+
+  generator_sm.update();
+
+  auto value = std::shared_ptr<const Signal>();
```
Comment (gustavoneves12) [inline tests/test_signal_generator.cpp]: You should check if the `value` is corrected.

**Comment by gustavoneves12** [comment]
@elizabetereis you need to improve the documentation and fix the indentation in your whole package. Remember, we use two space indentation.

File: include/hyro/SignalGeneratorComponent.h (hunk 11 above)
Comment (elizabetereis) [reply include/hyro/SignalGeneratorComponent.h]: Done. For all.

**Pull request merged**